toolchain/sem_ir/dump.cpp

```diff
@@ -0,0 +1,244 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef NDEBUG
+
+#include "toolchain/sem_ir/dump.h"
+
+#include "common/ostream.h"
+#include "toolchain/sem_ir/stringify_type.h"
+
+namespace Carbon::SemIR {
+
+static auto DumpIfValid(const File& file, NameId name_id) -> void {
+  if (name_id.is_valid()) {
+    llvm::errs() << " `" << file.names().GetFormatted(name_id) << "`";
+  }
+}
+
+static auto DumpNoNewline(const File& file, ConstantId const_id) -> void {
+  llvm::errs() << const_id;
+  if (const_id.is_symbolic()) {
+    llvm::errs() << ": "
+                 << file.constant_values().GetSymbolicConstant(const_id);
+  } else if (const_id.is_template()) {
+    llvm::errs() << ": "
+                 << file.insts().Get(
+                        file.constant_values().GetInstId(const_id));
+  }
+}
+
+static auto DumpNoNewline(const File& file, InstId inst_id) -> void {
+  llvm::errs() << inst_id;
+  if (inst_id.is_valid()) {
+    llvm::errs() << ": " << file.insts().Get(inst_id);
+  }
+}
+
+static auto DumpNoNewline(const File& file, InterfaceId interface_id) -> void {
+  llvm::errs() << interface_id;
+  if (interface_id.is_valid()) {
+    auto interface = file.interfaces().Get(interface_id);
+    llvm::errs() << ": " << interface;
+    DumpIfValid(file, interface.name_id);
+  }
+}
+
+static auto DumpNoNewline(const File& file, SpecificId specific_id) -> void {
+  llvm::errs() << specific_id;
+  if (specific_id.is_valid()) {
+    llvm::errs() << ": " << file.specifics().Get(specific_id);
+  }
+}
+
+LLVM_DUMP_METHOD auto Dump(const File& file, ClassId class_id) -> void {
+  llvm::errs() << class_id;
```

Comment:
The idea with the NoNewlines is that all the dumping logic lives in those functions so that they can be composed into other DumpNoNewline classes, and the body of each Dump() is just a DumpNoNewline() + newline. So this body here could move into a DumpNoNewline function.

Same for the other Dump functions below.

Reply:
I recognized that was the intent, and that is how I started writing this PR,  but in practice it didn't seem to work out that way. The information I wanted to print for a given id (a) benefited from formatting using newlines and indents and (b) didn't decompose into calls to `DumpNoNewline` functions. To avoid extra boilerplate, I've instead only written `DumpNoNewline` functions when there was a caller that wanted the functionality of one of the `Dump` functions without the newline, instead of writing an extra function proactively.